startup-scripts.md

```diff
@@ -0,0 +1,308 @@
+END
+
+# Create the VM spec
+cat << END > my-vm.yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VirtualMachine
+metadata:
+  name: sshvm
+spec:
+  terminationGracePeriodSeconds: 0
+  domain:
+    resources:
+      requests:
+        memory: 1024M
+    devices:
+      disks:
+      - name: registrydisk
+        volumeName: registryvolume
+        disk:
+          dev: vda
+      - name: cloudinitdisk
+        volumeName: cloudinitvolume
+        disk:
+          dev: vdb
+  volumes:
+    - name: registryvolume
+      registryDisk:
+        image: kubevirt/fedora-atomic-registry-disk-demo:devel
+    - name: cloudinitvolume
+      cloudInitNoCloud:
+        userDataBase64: $(cat startup-script | base64 -w0)
+END
+
+# Post the Virtual Machine spec to KubeVirt.
+kubectl create -f my-vm.yaml
+
+# Connect to VM with passwordless ssh key
+ssh -i <insert private key here> fedora@<insert ip here>
+
+```
+
+### Inject SSH key using Custom Shell Script
+
+Depending on the boot image in use, users may have a mixed experience using
+Cloud-Init's cloud-config to create users and inject ssh keys.
+
+Below is an example of creating a user and injecting ssh keys for that user
+using a script instead of cloud-config.
+
+```
+cat << END > startup-script.sh
+#!/bin/bash
+export NEW_USER="foo"
+export SSH_PUB_KEY="ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC6zdgFiLr1uAK7PdcchDd+LseA5fEOcxCCt7TLlr7Mx6h8jUg+G+8L9JBNZuDzTZSF0dR7qwzdBBQjorAnZTmY3BhsKcFr8Gt4KMGrS6r3DNmGruP8GORvegdWZuXgASKVpXeI7nCIjRJwAaK1x+eGHwAWO9Z8ohcboHbLyffOoSZDSIuk2kRIc47+ENRjg0T6x2VRsqX27g6j4DfPKQZGk0zvXkZaYtr1e2tZgqTBWqZUloMJK8miQq6MktCKAS4VtPk0k7teQX57OGwD6D7uo4b+Cl8aYAAwhn0hc0C2USfbuVHgq88ESo2/+NwV4SQcl3sxCW21yGIjAGt4Hy7J $NEW_USER@localhost.localdomain"
+
+sudo adduser -U -m $NEW_USER
+echo "$NEW_USER:atomic" | chpasswd
+sudo mkdir /home/$NEW_USER/.ssh
+sudo echo "$SSH_PUB_KEY" > /home/$NEW_USER/.ssh/authorized_keys
+sudo chown -R ${NEW_USER}: /home/$NEW_USER/.ssh
+END
+
+# Create the VM spec
+cat << END > my-vm.yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VirtualMachine
+metadata:
+  name: sshvm
+spec:
+  terminationGracePeriodSeconds: 0
+  domain:
+    resources:
+      requests:
+        memory: 1024M
+    devices:
+      disks:
+      - name: registrydisk
+        volumeName: registryvolume
+        disk:
+          dev: vda
+      - name: cloudinitdisk
+        volumeName: cloudinitvolume
+        disk:
+          dev: vdb
+  volumes:
+    - name: registryvolume
+      registryDisk:
+        image: kubevirt/fedora-atomic-registry-disk-demo:devel
+    - name: cloudinitvolume
+      cloudInitNoCloud:
+        userDataBase64: $(cat startup-script.sh | base64 -w0)
+END
+
+# Post the Virtual Machine spec to KubeVirt.
```

Comment:
In this example, the startup script at around line 248 was never put into a secret. Anybody following along with the entire document would know to do that, but anybody that skipped straight to this example might miss that point.

Reply:
It's not required to put the script in a k8s secret. There's two ways supplying the data. 
- k8s secret referenced in userDataSecretRef field on volume spec
- base64 encoded script copied directly into the userDataBase64 field on the volume spec.

I gave an example of the k8s secret method somewhere in the middle of the doc, but for the rest of the examples i'm just using the userDataBase64 method.

Reply:
+1